### DISCUSSION TITLE
Creating values at non-existing paths with json_pointer: is it an undocumented behaviour?

### DISCUSSION THREAD

**Discussion by shubhampaul** [nlohmann/json#4446] [General]
I am using v3.11.3

```
json j;
json::json_pointer ptr("/foo/0/0/0");
j[ptr] = "12";
std::cout << std::setw(4) << j << std::endl;
```

For the above ptr, using operator[] to insert elements into a non-exisitng nested path, what is to be expected? It creates the below json:

```
 {
     "foo": [
         [
             [
                 "12"
             ]
        ]
    ]
 }

```

So it seems to interpret the non-numeric token as a key to create and insert an object where nothing existed previously and then numeric tokens are interpreted as array indices which creates arrays for the previously non-existing positions. 

If I change the pointer to `/foo/one/one/one`, it creates the following:

```
  {
     "foo": {
         "one": {
             "one": {
                 "one": "12"
             }
         }
     }
 }
```

If I understand it correctly then for non-existing values, the type (object or array) is inserted depending on if the token in the json pointer is convertible to a number or not? So if it is convertible to a number then, it is treated like an array index which creates an array otherwise it is treated like an object key and an object is inserted?

The documentation mentions how numeric and non-numeric token inserts elements in a pre-existing type (either array or object) but doesn't talk (or maybe I can't find it?) about how it behaves when nothing exists. Is this undocumented behaviour or do I not understand the documentation? Would really appreciate if someone could clarify this.

**Comment by nlohmann** [comment]
> If I understand it correctly then for non-existing values, the type (object or array) is inserted depending on if the token in the json pointer is convertible to a number or not? So if it is convertible to a number then, it is treated like an array index which creates an array otherwise it is treated like an object key and an object is inserted?

Yes, this is the behavior. The RFC for JSON Pointer does not specify how values are created, so we had to improvise here.

> The documentation mentions how numeric and non-numeric token inserts elements in a pre-existing type (either array or object) but doesn't talk (or maybe I can't find it?) about how it behaves when nothing exists. Is this undocumented behaviour or do I not understand the documentation? Would really appreciate if someone could clarify this.

https://json.nlohmann.me/api/basic_json/operator%5B%5D/ is the documentation, and it does not specifically mentions the strategy for JSON Pointers. We mimic the sam behavior for `operator[]` for numbers and strings - the former are treated as array indices and the latter as object keys.
